infrastructure/src/main/java/org/corfudb/infrastructure/management/FileSystemAdvisor.java

```diff
@@ -0,0 +1,73 @@
+package org.corfudb.infrastructure.management;
+
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.protocols.wireprotocol.ClusterState;
+import org.corfudb.protocols.wireprotocol.NodeState;
+import org.corfudb.protocols.wireprotocol.failuredetector.FileSystemStats;
+import org.corfudb.protocols.wireprotocol.failuredetector.NodeRank.NodeRankByResourceQuota;
+
+import java.util.NavigableSet;
+import java.util.Optional;
+import java.util.TreeSet;
+
+/**
+ * Detects failures in the file system
+ */
+@Slf4j
+public class FileSystemAdvisor {
+
+    /**
+     * Checks if a node is failed because of quota exceeded
+     *
+     * @param clusterState cluste state
+     * @return node rank by quota
+     */
+    public Optional<NodeRankByResourceQuota> findFailedNodeByResourceQuota(ClusterState clusterState) {
+        //find a node which has no resource quota
+        NavigableSet<NodeRankByResourceQuota> set = new TreeSet<>();
+        for (NodeState node : clusterState.getNodes().values()) {
+            node.getFileSystem().ifPresent(fsStats -> {
+                if (!fsStats.getQuota().isExceeded()) {
+                    return;
+                }
+
+                NodeRankByResourceQuota quota = new NodeRankByResourceQuota(
+                        node.getConnectivity().getEndpoint(),
+                        fsStats.getQuota()
+                );
+                set.add(quota);
+            });
+        }
+
+        return Optional.ofNullable(set.pollLast());
```

Comment:
This is what @zfrenette pointed out, I don't think we should mark nodes with QuotaExceeded as unresponsive. It doesn't improve the availability because all nodes have to replicate the same set of data.

Reply:
Sure, I'll change this piece of code in the 3-rd pr, as we disccussed